q2_mlab/plotting/app.py

```diff
@@ -0,0 +1,613 @@
+        where = keep_repeats[across] == sub_group
+        keep_repeats_parts.append(keep_repeats.loc[where])
+        keep_repeats_parts[i].columns = [col + '_' + sub_group for
+                                         col in keep_repeats_parts[i].columns]
+
+    segment_df = pd.concat(keep_repeats_parts,
+                           axis=1
+                           )
+    return segment_df
+
+
+class TextInputComponent(ComponentMixin):
+    def __init__(self, text_input_kwargs):
+        super().__init__()
+        self.text_input = TextInput(**text_input_kwargs)
+        self.layout = self.text_input
+        self.input_text_callback = None
+
+    def set_mediator(self, mediator):
+        super().set_mediator(mediator)
+        event_name = 'text-change'
+        text_change = self.make_attr_old_new_callback(event_name)
+        self.input_text_callback = text_change
+        self.text_input.on_change('value', self.input_text_callback)
+
+
+class CheckboxButtonGroupComponent(ComponentMixin):
+    def __init__(self, checkbox_kwargs):
+        super().__init__()
+        self.checkbox = CheckboxButtonGroup(**checkbox_kwargs)
+        self.checkbox_change = None
+        self.layout = self.checkbox
+
+    def set_mediator(self, mediator):
+        super().set_mediator(mediator)
+        event_name = 'checkbox-change'
+        self.checkbox_change = self.make_attr_old_new_callback(event_name)
+        self.checkbox.on_change('active', self.checkbox_change)
+
+
+class SegmentComponentExt(SegmentComponent):
+    def redraw(self, x, y, seg_0, seg_1, data):
+        self.data_source.data = data
+        self.segment.glyph.x0 = '_'.join([x, seg_0])
+        self.segment.glyph.x1 = '_'.join([x, seg_1])
+        self.segment.glyph.y0 = '_'.join([y, seg_0])
+        self.segment.glyph.y1 = '_'.join([y, seg_1])
+
+
+palettes = {
+    'Category20': Category20,
+    'Set3': Set3,
+}
+
+
+DEFAULTS = {
+    'segment_variable': 'dataset',
+    'x': 'MAE_mean',
+    'y': 'MAE_var',
+    'x_axis_type': 'log',
+    'y_axis_type': 'log',
+    'cmap': 'Category20'
+}
+
+
+class AlgorithmScatter(Mediator, Plottable):
+
+    def __init__(self, x, y, engine, cmap=None):
+        super().__init__()
+        self.x = x
+        self.y = y
+        self.engine = engine
+        self.data = None
+        self.scatter = None
+        if cmap is None:
+            self.cmap = Category20
+        else:
+            self.cmap = cmap
+        self.line_segment_variable = DEFAULTS['segment_variable']
+        self.data_raw = None
+        self.data_static = None
+        self.data = None
+        self.seg_0, self.seg_1 = None, None
+        self.scatter_source = None
+        self.x_axis_type = DEFAULTS['x_axis_type']
+        self.y_axis_type = DEFAULTS['y_axis_type']
+        self.axis_types = ['linear', 'log']
+        self.line_segment_pairs = {
+            'dataset': ['finrisk', 'sol'],
+            'level': ['16S', 'MG'],
+            'target': ['age', 'bmi'],
```

Comment:
Should we expose these to the user since we'll be looking at more than just these datasets? Or do you want to save that for a future PR?

Reply:
Probably better for a future PR.